### PR TITLE
Warn SDPA users about dropout behavior

### LINKED ISSUE #124464: SDPA does not switch off dropout during evaluation
### 🐛 Describe the bug

```
# Efficient implementation equivalent to the following:
def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False, scale=None) -> torch.Tensor:
    # Efficient implementation equivalent to the following:
    L, S = query.size(-2), key.size(-2)
    scale_factor = 1 / math.sqrt(query.size(-1)) if scale is None else scale
    attn_bias = torch.zeros(L, S, dtype=query.dtype)
    if is_causal:
        assert attn_mask is None
        temp_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
        attn_bias.masked_fill_(temp_mask.logical_not(), float("-inf"))
        attn_bias.to(query.dtype)

    if attn_mask is not None:
        if attn_mask.dtype == torch.bool:
            attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
        else:
            attn_bias += attn_mask
    attn_weight = query @ key.transpose(-2, -1) * scale_factor
    attn_weight += attn_bias
    attn_weight = torch.softmax(attn_weight, dim=-1)
    attn_weight = torch.dropout(attn_weight, dropout_p, train=True) # Bug Here!!
    return attn_weight @ value 
```

I find out the detailed pseudo-like code from https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html while using flash attention.

However the metric measured from flash attention model is changed (especially when I use dropout_p = 0.1) which is supposed to be unchanged. Flash attention does not affect the output as I know. Then I figure out the expected bug from the code above and marked with annotation "Bug Here!!".

I execute the two code blocks following:

```
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import random

torch.manual_seed(42)
torch.cuda.manual_seed(42)
np.random.seed(42)
random.seed(42)
torch.backends.cudnn.deterministic = True

# Using functional dropout
class Model1(nn.Module):
    def __init__(self, p=0.5):
        super().__init__()
        self.p = p

    def forward(self, inputs):
        return F.dropout(inputs, p=self.p, training=True)

# Example usage
model1 = Model1(p=0.5)

inputs = torch.rand(10)
print("Random input", inputs)
print()

# Applying dropout in training mode
print('Training mode:')
print('Model 1', model1(inputs))

# Switching to evaluation mode
model1.eval()

# Applying dropout in evaluation mode
print('Evaluation mode:')
print('Model 1', model1(inputs))
```

and

```
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import random

torch.manual_seed(42)
torch.cuda.manual_seed(42)
np.random.seed(42)
random.seed(42)
torch.backends.cudnn.deterministic = True

# Using nn.Dropout module
class Model2(nn.Module):
    def __init__(self, p=0.5):
        super().__init__()
        self.drop_layer = nn.Dropout(p=p)

    def forward(self, inputs):
        return self.drop_layer(inputs)
# Example usage
model2 = Model2(p=0.5)

inputs = torch.rand(10)
print("Random input", inputs)
print()

# Applying dropout in training mode
print('Training mode:')
print('Model 2', model2(inputs))

# Switching to evaluation mode
model2.eval()

# Applying dropout in evaluation mode
print('Evaluation mode:')
print('Model 2', model2(inputs))
```

The result for them is following:

스크린샷 2024-04-19 오후 5 04 13

스크린샷 2024-04-19 오후 5 04 23

As you look at the model 1, the same process as dropout in scaled_dot_product_attention, it does not act correctly even though I turn on the evaluation option. So please fix this one as soon as possible. Thank you.

### Versions

Collecting environment information...
PyTorch version: 2.3.0.dev20240121+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-159-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.0.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090
GPU 3: NVIDIA GeForce RTX 3090

Nvidia driver version: 535.154.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             80
On-line CPU(s) list:                0-79
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 5218R CPU @ 2.10GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 2
Core(s) per socket:                 20
Socket(s):                          2
Stepping:                           7
CPU max MHz:                        4000.0000
CPU min MHz:                        800.0000
BogoMIPS:                           4200.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          1.3 MiB (40 instances)
L1i cache:                          1.3 MiB (40 instances)
L2 cache:                           40 MiB (40 instances)
L3 cache:                           55 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78
NUMA node1 CPU(s):                  1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: Split huge pages
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.1
[pip3] pytorch-lightning==2.1.2
[pip3] pytorch-triton==2.2.0+e28a256d71
[pip3] torch==2.3.0.dev20240121+cu121
[pip3] torch-tb-profiler==0.4.3
[pip3] torchaudio==2.2.0.dev20240121+cu121
[pip3] torchmetrics==1.3.0.post0
[pip3] torchvision==0.18.0.dev20240121+cu121
[conda] Could not collect

cc @svekars @brycebortree @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126294

Fixes #124464

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -5016,6 +5016,24 @@ def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.
 
 .. warning:: This function is beta and subject to change.
 
+.. warning::
+
+    This function always applies dropout according to the specified `dropout_p` argument.
+    To disable dropout during evaluation, be sure to pass a value of `0.0` when the module
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:5022]: nit: 
```suggestion
    This function always applies dropout according to the specified ``dropout_p`` argument.
    To disable dropout during evaluation, be sure to pass a value of ``0.0`` when the module
```

**Review by mikaylagawarecki — APPROVED** [review]
Rendering looks good other than the nit

**Review by drisspg — APPROVED** [review]
